bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/CommunicationManager.java

```diff
@@ -427,118 +423,37 @@ private <T extends Type> void handleEvent(String itemName, T type, @Nullable Str
-            if (quantityType != null) {
-                return (T) quantityType;
-            }
-        }
+        String channelAcceptedItemType = channel.getAcceptedItemType();
 
-        // The command is sent to an item w/o dimension defined and the channel is legacy (created from a ThingType
-        // definition before UoM was introduced to the binding). The dimension information might now be defined on the
-        // current ThingType. The binding might expect us to provide a QuantityType so try to convert to the dimension
-        // the ChannelType provides.
-        // This can be removed once a suitable solution for https://github.com/eclipse/smarthome/issues/2555 (Thing
-        // migration) is found.
-        if (originalType instanceof DecimalType type && !hasDimension(item, acceptedItemType)
-                && channelTypeDefinesDimension(channel.getChannelTypeUID())) {
-            ChannelType channelType = channelTypeRegistry.getChannelType(channel.getChannelTypeUID());
-
-            String acceptedItemTypeFromChannelType = channelType != null ? channelType.getItemType() : null;
-            @Nullable
-            QuantityType<?> quantityType = convertToQuantityType(type, item, acceptedItemTypeFromChannelType);
-            if (quantityType != null) {
-                return (T) quantityType;
-            }
-        }
-
-        if (acceptedItemType == null) {
+        if (channelAcceptedItemType == null) {
             return originalType;
         }
 
-        List<Class<? extends T>> acceptedTypes = acceptedTypesFunction.apply(acceptedItemType);
-        if (acceptedTypes == null) {
-            return originalType;
+        // handle Number-Channels for backward compatibility
+        if ("Number".equals(channelAcceptedItemType) && originalType instanceof QuantityType<?> quantityType) {
+            // strip unit from QuantityType for channels that accept plain number
+            return (T) new DecimalType(quantityType.toBigDecimal());
         }
 
-        if (acceptedTypes.contains(originalType.getClass())) {
-            return originalType;
-        } else {
-            // Look for class hierarchy and convert appropriately
-            for (Class<? extends T> typeClass : acceptedTypes) {
-                if (!typeClass.isEnum() && typeClass.isAssignableFrom(originalType.getClass()) //
-                        && State.class.isAssignableFrom(typeClass) && originalType instanceof State state) {
-                    T ret = (T) state.as((Class<? extends State>) typeClass);
-                    if (logger.isDebugEnabled()) {
-                        logger.debug("Converted '{}' ({}) to accepted type '{}' ({}) for channel '{}' ", originalType,
-                                originalType.getClass().getSimpleName(), ret, ret.getClass().getName(),
-                                channel.getUID());
-                    }
-                    return ret;
-                }
-            }
```

Comment:
This is interesting. I removed the code and replaced it by the `Dimmer/HSBType` check, because it did only affect the return type in two cases:

```
[ERROR] Failures: 
[ERROR]   CommunicationManagerConversionTest.testCommand:114 CommunicationManager did a conversion for target item Rollershutter from class org.openhab.core.library.types.HSBType to class org.openhab.core.library.types.PercentType
[ERROR]   CommunicationManagerConversionTest.testCommand:114 CommunicationManager did a conversion for target item Dimmer from class org.openhab.core.library.types.HSBType to class org.openhab.core.library.types.PercentType    
```

One is covered by the simple check that I added and the other one is IMO wrong. Sending `HSBType` to a `RollershutterItem` doesn't make any sense.

Reply:
Yes, sending `HSBType` to rollershutters does not need to be supported.
The check seems to cover hierarchical types in a very generic way, but since we are unlikely to add further hierarchical types in future, I agree that this complex check isn't needed here.